diff -up openvswitch-2.3.90-10031.gitf097013a/acinclude.m4.dpdk-lib openvswitch-2.3.90-10031.gitf097013a/acinclude.m4
--- openvswitch-2.3.90-10031.gitf097013a/acinclude.m4.dpdk-lib	2015-04-16 09:08:48.000000000 +0300
+++ openvswitch-2.3.90-10031.gitf097013a/acinclude.m4	2015-04-17 16:20:51.163345523 +0300
@@ -170,12 +170,11 @@ AC_DEFUN([OVS_CHECK_DPDK], [
 
     DPDK_INCLUDE=$RTE_SDK/include
     DPDK_LIB_DIR=$RTE_SDK/lib
-    DPDK_LIB="-lintel_dpdk"
-    DPDK_EXTRA_LIB="-lfuse"
+    DPDK_LIB="-ldpdk_core"
 
     ovs_save_CFLAGS="$CFLAGS"
     ovs_save_LDFLAGS="$LDFLAGS"
-    LDFLAGS="$LDFLAGS -L$DPDK_LIB_DIR"
+    LDFLAGS="$LDFLAGS -L."
     CFLAGS="$CFLAGS -I$DPDK_INCLUDE"
 
     # On some systems we have to add -ldl to link with dpdk
@@ -204,7 +203,7 @@ AC_DEFUN([OVS_CHECK_DPDK], [
     fi
     CFLAGS="$ovs_save_CFLAGS"
     LDFLAGS="$ovs_save_LDFLAGS"
-    OVS_LDFLAGS="$OVS_LDFLAGS -L$DPDK_LIB_DIR"
+    OVS_LDFLAGS="$OVS_LDFLAGS -L."
     OVS_CFLAGS="$OVS_CFLAGS -I$DPDK_INCLUDE"
 
     # DPDK pmd drivers are not linked unless --whole-archive is used.
