commit edb71e44bbe788a46febcd94d8ce0a6abc674c98
Author: Panu Matilainen <pmatilai@redhat.com>
Date:   Mon Apr 4 14:16:50 2016 +0300

    netdev-dpdk: adjust to DPDK 16.04 ethdev speed API

diff --git a/lib/netdev-dpdk.c b/lib/netdev-dpdk.c
index 4cdec3c..f48fe41 100644
--- a/lib/netdev-dpdk.c
+++ b/lib/netdev-dpdk.c
@@ -1573,31 +1573,29 @@ netdev_dpdk_get_features(const struct netdev *netdev_,
     link = dev->link;
     ovs_mutex_unlock(&dev->mutex);
 
-    if (link.link_duplex == ETH_LINK_AUTONEG_DUPLEX) {
-        if (link.link_speed == ETH_LINK_SPEED_AUTONEG) {
-            *current = NETDEV_F_AUTONEG;
-        }
+    if (link.link_autoneg) {
+	*current = NETDEV_F_AUTONEG;
     } else if (link.link_duplex == ETH_LINK_HALF_DUPLEX) {
-        if (link.link_speed == ETH_LINK_SPEED_10) {
+        if (link.link_speed == ETH_LINK_SPEED_10M) {
             *current = NETDEV_F_10MB_HD;
         }
-        if (link.link_speed == ETH_LINK_SPEED_100) {
+        if (link.link_speed == ETH_LINK_SPEED_100M) {
             *current = NETDEV_F_100MB_HD;
         }
-        if (link.link_speed == ETH_LINK_SPEED_1000) {
+        if (link.link_speed == ETH_LINK_SPEED_1G) {
             *current = NETDEV_F_1GB_HD;
         }
     } else if (link.link_duplex == ETH_LINK_FULL_DUPLEX) {
-        if (link.link_speed == ETH_LINK_SPEED_10) {
+        if (link.link_speed == ETH_LINK_SPEED_10M) {
             *current = NETDEV_F_10MB_FD;
         }
-        if (link.link_speed == ETH_LINK_SPEED_100) {
+        if (link.link_speed == ETH_LINK_SPEED_100M) {
             *current = NETDEV_F_100MB_FD;
         }
-        if (link.link_speed == ETH_LINK_SPEED_1000) {
+        if (link.link_speed == ETH_LINK_SPEED_1G) {
             *current = NETDEV_F_1GB_FD;
         }
-        if (link.link_speed == ETH_LINK_SPEED_10000) {
+        if (link.link_speed == ETH_LINK_SPEED_10G) {
             *current = NETDEV_F_10GB_FD;
         }
     }
