commit adc2ef9ec27dd7fa2502c55a3c12da96910d0e5e
Author: Panu Matilainen <pmatilai@redhat.com>
Date:   Mon Mar 23 14:42:32 2015 +0200

    Support DPDK 2.0

diff --git a/lib/netdev-dpdk.c b/lib/netdev-dpdk.c
index 4e16f39..ee98ba0 100644
--- a/lib/netdev-dpdk.c
+++ b/lib/netdev-dpdk.c
@@ -117,8 +117,12 @@ static const struct rte_eth_conf port_conf = {
     .rx_adv_conf = {
         .rss_conf = {
             .rss_key = NULL,
+#ifdef ETH_RSS_IPV4_TCP /* dpdk < 2.0 */
             .rss_hf = ETH_RSS_IPV4_TCP | ETH_RSS_IPV4 | ETH_RSS_IPV6
                     | ETH_RSS_IPV4_UDP | ETH_RSS_IPV6_TCP | ETH_RSS_IPV6_UDP,
+#else
+            .rss_hf = ETH_RSS_IP | ETH_RSS_UDP | ETH_RSS_TCP,
+#endif
         },
     },
     .txmode = {
