diff -up openvswitch-2.4.90-11130.git3e2493e1/acinclude.m4.dpdk-lib openvswitch-2.4.90-11130.git3e2493e1/acinclude.m4
--- openvswitch-2.4.90-11130.git3e2493e1/acinclude.m4.dpdk-lib	2015-11-13 00:33:03.000000000 +0200
+++ openvswitch-2.4.90-11130.git3e2493e1/acinclude.m4	2015-11-13 09:23:04.937690944 +0200
@@ -218,16 +218,6 @@ AC_DEFUN([OVS_CHECK_DPDK], [
     OVS_LDFLAGS="$OVS_LDFLAGS -L$DPDK_LIB_DIR"
     OVS_CFLAGS="$OVS_CFLAGS -I$DPDK_INCLUDE -mssse3"
 
-    # DPDK pmd drivers are not linked unless --whole-archive is used.
-    #
-    # This happens because the rest of the DPDK code doesn't use any symbol in
-    # the pmd driver objects, and the drivers register themselves using an
-    # __attribute__((constructor)) function.
-    #
-    # These options are specified inside a single -Wl directive to prevent
-    # autotools from reordering them.
-    DPDK_vswitchd_LDFLAGS=-Wl,--whole-archive,$DPDK_LIB,--no-whole-archive
-    AC_SUBST([DPDK_vswitchd_LDFLAGS])
     AC_DEFINE([DPDK_NETDEV], [1], [System uses the DPDK module.])
 
   else
