diff -up openvswitch-2.4.90-10916.gita2cf7524/acinclude.m4.dpdk-lib openvswitch-2.4.90-10916.gita2cf7524/acinclude.m4
--- openvswitch-2.4.90-10916.gita2cf7524/acinclude.m4.dpdk-lib	2015-09-28 17:34:37.006842399 +0300
+++ openvswitch-2.4.90-10916.gita2cf7524/acinclude.m4	2015-09-28 17:35:35.554697359 +0300
@@ -171,7 +171,7 @@ AC_DEFUN([OVS_CHECK_DPDK], [
     RTE_SDK=$with_dpdk
 
     DPDK_INCLUDE=$RTE_SDK/include
-    DPDK_LIB_DIR=$RTE_SDK/lib
+    DPDK_LIB_DIR="."
     DPDK_LIB="-ldpdk"
     DPDK_EXTRA_LIB=""
     RTE_SDK_FULL=`readlink -f $RTE_SDK`
