diff -up openvswitch-2.3.90-9961.gitb3cceba0/lib/netdev-dpdk.c.dpdk-rxburst openvswitch-2.3.90-9961.gitb3cceba0/lib/netdev-dpdk.c
--- openvswitch-2.3.90-9961.gitb3cceba0/lib/netdev-dpdk.c.dpdk-rxburst	2015-04-09 13:23:23.952357811 +0300
+++ openvswitch-2.3.90-9961.gitb3cceba0/lib/netdev-dpdk.c	2015-04-09 13:23:44.661307435 +0300
@@ -149,7 +149,7 @@ static const struct rte_eth_txconf tx_co
     .txq_flags = ETH_TXQ_FLAGS_NOMULTSEGS|ETH_TXQ_FLAGS_NOOFFLOADS,
 };
 
-enum { MAX_RX_QUEUE_LEN = 192 };
+enum { MAX_RX_QUEUE_LEN = 32 };
 enum { MAX_TX_QUEUE_LEN = 384 };
 enum { DPDK_RING_SIZE = 256 };
 BUILD_ASSERT_DECL(IS_POW2(DPDK_RING_SIZE));
